File: examples/e-ktor-example/src/dev/akif/ektorexample/common/Repository.kt
package dev.akif.ektorexample.common

import dev.akif.ektorexample.database.DB
import e.kotlin.*
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.transactions.transaction

abstract class Repository<M>(open val db: DB) {
    abstract fun convertTo(row: ResultRow): M

    fun <T> run(block: () -> T): EOr<T> =
        EOr.catching({ transaction { block() } }) { t ->
            Errors.database.cause(t.toE())
        }
}

File: examples/e-play-example/app/dev/akif/eplayexample/common/Repository.scala
package dev.akif.eplayexample.common

import java.sql.Connection

import e.ezio._
import e.scala._

trait Repository {
    val db: DB.Def

    protected def run[A](f: Connection => A): EIO[A] =
        db.withConnection(f)
            .catching(t => Errors.database.message("Database error!").cause(t.toE()))
            .toEIO
}
